### PR TITLE
Double zip FlutterMacOS.dSYM.zip.

### PR DESCRIPTION
**Pull request by godofredoc**
The flutter tool is expecting FlutterMacOS.dSYM.zip to be double zipped.

Bug: https://github.com/flutter/flutter/issues/124911

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by godofredoc** [comment]
\cc @XilaiZhang this will impact codesigning in two different parts: identifying if the file requires signing and also on the signing tool detecting if it needs to unzip twice.

**Comment by XilaiZhang** [comment]
> \cc @XilaiZhang this will impact codesigning in two different parts: identifying if the file requires signing and also on the signing tool detecting if it needs to unzip twice.

thanks for the info! Yeah I think the current topology covers this case.

based on the insights from https://github.com/flutter/flutter/issues/124322#issuecomment-1499409762 we would skip signing this file since it is debug symbols and not executables. On the signing tool side, this file will be skipped since if no signing metadata is found upon the first unzip, the file will be skipped.

File: sky/tools/create_macos_framework.py [hunk 1]
```diff
@@ -152,26 +164,29 @@ def process_framework(dst, args, fat_framework, fat_framework_binary):
...
     )
     subprocess.check_call([
         'zip',
         '-r',
         '-y',
         'FlutterMacOS.framework.zip',
-        'FlutterMacOS.framework',
+        '.',
         'entitlements.txt',
         'without_entitlements.txt',
```
Comment (XilaiZhang) [inline sky/tools/create_macos_framework.py:187]: Right we would remove the metadata from the inner layer of the zip, and move them to the outer layer of the zip, so we will remove L186-187.

File: sky/tools/create_macos_framework.py [hunk 2]
```diff
@@ -180,9 +195,9 @@ def process_framework(dst, args, fat_framework, fat_framework_binary):
         'FlutterMacOS.framework_.zip',
         'FlutterMacOS.framework.zip',
     ],
```
Comment (XilaiZhang) [inline sky/tools/create_macos_framework.py:197]: yeah we would add codesign metadata here in the outer layer of the zip, something similar to  
         'zip',
        '-y',
        'FlutterMacOS.framework_.zip',
        'FlutterMacOS.framework.zip',
        'entitlements.txt',
        'without_entitlements.txt',

So adding  'entitlements.txt', 'without_entitlements.txt', on L197 and L198

File: sky/tools/create_macos_framework.py [hunk 3]
```diff
@@ -152,26 +164,29 @@ def process_framework(dst, args, fat_framework, fat_framework_binary):
   # Zip FlutterMacOS.framework.
   if args.zip:
     filepath_with_entitlements = ''
-    filepath_without_entitlements = 'FlutterMacOS.framework/Versions/A/FlutterMacOS'
+
+    framework_dst = os.path.join(dst, 'FlutterMacOS.framework')
+    filepath_without_entitlements = 'Versions/A/FlutterMacOS'
```
Comment (XilaiZhang) [inline sky/tools/create_macos_framework.py:169]: The metadata path will now be `FlutterMacOS.framework.zip/Versions/A/FlutterMacOS`
so L169 will be like `filepath_without_entitlements = FlutterMacOS.framework.zip/Versions/A/FlutterMacOS`

**Review by XilaiZhang — COMMENTED** [review]
Thanks for explaining! I added the changes needed for codesign metadata in comments.

I tested the correctness of the new change by downloading the artifacts built from bab6f0e811596da498e70372c3fc91ca03ebb44c engine v2 bucket, and run codesign locally on it.

File: sky/tools/create_macos_framework.py (hunk 1 above)
Comment (godofredoc) [reply sky/tools/create_macos_framework.py:187]: duplicated the metadata to simplify removing the outer zip later.

File: sky/tools/create_macos_framework.py (hunk 2 above)
Comment (godofredoc) [reply sky/tools/create_macos_framework.py:197]: ack

File: sky/tools/create_macos_framework.py (hunk 3 above)
Comment (godofredoc) [reply sky/tools/create_macos_framework.py:169]: ack

Comment (godofredoc) [reply sky/tools/create_macos_framework.py:169]: Actually adding this seems overcomplicated. As a workaround I'll check the file name explicitly in the recipes with a TODO to remove it after the outer zip is deleted.

File: sky/tools/create_macos_framework.py (hunk 1 above)
Comment (godofredoc) [reply sky/tools/create_macos_framework.py:187]: That will complicate the logic in several places. As the outer zipped is expected to be removed ~2weeks from now I'll hard code the name in recipes for now.

**Comment by godofredoc** [comment]
> > \cc @XilaiZhang this will impact codesigning in two different parts: identifying if the file requires signing and also on the signing tool detecting if it needs to unzip twice.
> 
> thanks for the info! Yeah I think the current topology covers this case.
> 
> based on the insights from [flutter/flutter#124322 (comment)](https://github.com/flutter/flutter/issues/124322#issuecomment-1499409762) we would skip signing this file since it is debug symbols and not executables. On the signing tool side, this file will be skipped since if no signing metadata is found upon the first unzip, the file will be skipped.

This is covered at the recipes level: https://flutter-review.googlesource.com/c/recipes/+/42560

**Comment by godofredoc** [comment]
> Thanks for explaining! I added the changes needed for codesign metadata in comments.
> 
> I tested the correctness of the new change by downloading the artifacts built from [bab6f0e](https://github.com/flutter/engine/commit/bab6f0e811596da498e70372c3fc91ca03ebb44c) engine v2 bucket, and run codesign locally on it.

ACK

**Comment by XilaiZhang** [comment]
> > > \cc @XilaiZhang this will impact codesigning in two different parts: identifying if the file requires signing and also on the signing tool detecting if it needs to unzip twice.
> > 
> > 
> > thanks for the info! Yeah I think the current topology covers this case.
> > based on the insights from [flutter/flutter#124322 (comment)](https://github.com/flutter/flutter/issues/124322#issuecomment-1499409762) we would skip signing this file since it is debug symbols and not executables. On the signing tool side, this file will be skipped since if no signing metadata is found upon the first unzip, the file will be skipped.
> 
> This is covered at the recipes level: https://flutter-review.googlesource.com/c/recipes/+/42560

Umm I left a comment in recipe. Sorry if I misunderstood, umm I think in the recipe we are requiring the zip to be codesigned, but we are supplying empty configs. My understanding is that the codesign tool might detect the binary in the zip and report a failure due to the empty config. Or maybe we are planning to add more commits in this pr

**Comment by godofredoc** [comment]
> > > > \cc @XilaiZhang this will impact codesigning in two different parts: identifying if the file requires signing and also on the signing tool detecting if it needs to unzip twice.
> > > 
> > > 
> > > thanks for the info! Yeah I think the current topology covers this case.
> > > based on the insights from [flutter/flutter#124322 (comment)](https://github.com/flutter/flutter/issues/124322#issuecomment-1499409762) we would skip signing this file since it is debug symbols and not executables. On the signing tool side, this file will be skipped since if no signing metadata is found upon the first unzip, the file will be skipped.
> > 
> > 
> > This is covered at the recipes level: https://flutter-review.googlesource.com/c/recipes/+/42560
> 
> Umm I left a comment in recipe. Sorry if I misunderstood, umm I think in the recipe we are requiring the zip to be codesigned, but we are supplying empty configs. My understanding is that the codesign tool might detect the binary in the zip and report a failure due to the empty config. Or maybe we are planning to add more commits in this pr

Details:

* The double zip behavior is temporary, ~couple of weeks until it is fixed.
* Having the inner zip with the correct structure(metadata with correct paths) would have been the easiest way to remove the outer zip, preventing additional disruption when the outer zip is removed.

**Comment by godofredoc** [comment]
@XilaiZhang implemented your suggestions and added to TODOs to follow up when removing the outer zip folders.

@christopherfujino @jmagman  any concerns about adding the double zip logic to the dsym artifact?

**Comment by jmagman** [comment]
> any concerns about adding the double zip logic to the dsym artifact?

I don't think the tool knows anything about the dSYM artifact, I don't think it downloads it at all.  Users manually download it https://github.com/flutter/flutter/wiki/Crashes (and macOS isn't mentioned).  Was that artifact actually causing issues somewhere?  (I know the .framework was https://github.com/flutter/flutter/issues/124911#issuecomment-1513263342).

**Comment by godofredoc** [comment]
> > any concerns about adding the double zip logic to the dsym artifact?
> 
> I don't think the tool knows anything about the dSYM artifact, I don't think it downloads it at all. Users manually download it https://github.com/flutter/flutter/wiki/Crashes (and macOS isn't mentioned). Was that artifact actually causing issues somewhere? (I know the .framework was [flutter/flutter#124911 (comment)](https://github.com/flutter/flutter/issues/124911#issuecomment-1513263342)).

It was not causing any issues but when checking all the artifacts I found the legacy dsym artifact is double zipped. I'm trying to get all the artifacts generated by engine v2 aligned with the legacy artifacts to avoid new disruptions during the migration.

**Review by XilaiZhang — APPROVED** [review]
LGTM! sorry for blocking v2 migration

As a side note, we could keep the metadata for both the inner zip and outer zip (have them both present), and remove the outer metadata when the outer zip is removed. This would work because the codesign tool only looks at metadata at the outer most layer.

**Pull request merged**